pysindy/optimizers/sbr.py

```diff
@@ -0,0 +1,163 @@
+        prior. See Hirsh et al. (2021) for more details
+
+    slab_shape_s : float, optional (default 2)
+        Shape parameter for the slab region of the regularised horseshoe
+        prior. See Hirsh et al. (2021) for more details
+
+    noise_hyper_lambda : float, optional (default 1)
+        Rate hyperparameter for the exponential prior distribution over
+        the noise standard deviation.
+
+    normalize_columns : boolean, optional (default False)
+        Normalize the columns of x (the SINDy library terms) before regression
+        by dividing by the L2-norm. Note that the 'normalize' option in sklearn
+        is deprecated in sklearn versions >= 1.0 and will be removed.
+
+    copy_X : boolean, optional (default True)
+        If True, X will be copied; else, it may be overwritten.
+
+    **mcmc_kwargs
+        Instructions for MCMC sampling.
+        Keyword arguments are passed to numpyro.infer.MCMC
+
+    Attributes
+    ----------
+    coef_ : array, shape (n_features,) or (n_targets, n_features)
+        Posterior means of the SINDy coefficients.
+
+    mcmc : numpyro.infer.MCMC
+        Complete traces of the posterior samples.
+    """
+
+    def __init__(
+        self,
+        sparsity_coef_tau0=0.1,
+        slab_shape_nu=4,
+        slab_shape_s=2,
+        noise_hyper_lambda=1,
+        normalize_columns=False,
+        copy_X=True,
+        **mcmc_kwargs,
+    ):
+        super().__init__(
+            copy_X=copy_X,
+            normalize_columns=normalize_columns,
+        )
+        # set the hyperparameters
+        self.sparsity_coef_tau0 = sparsity_coef_tau0
+        self.slab_shape_nu = slab_shape_nu
+        self.slab_shape_s = slab_shape_s
+        self.noise_hyper_lambda = noise_hyper_lambda
+
+        # store the MCMC kwargs.
+        self.mcmc_kwargs = mcmc_kwargs
+
+    def _reduce(self, x, y):
+        # set up a sparse regression and sample.
+        self.mcmc = self._run_mcmc(x, y, **self.mcmc_kwargs)
+
+        # set the mean values as the coefficients.
+        self.coef_ = np.array(self.mcmc.get_samples()["beta"].mean(axis=0))
+
+    def _numpyro_model(self, x, y):
+        # get the dimensionality of the problem.
+        n_features = x.shape[1]
+        n_targets = y.shape[1]
+
+        # sample the hyperparameters.
+        tau, c_sq = sample_reg_horseshoe_hyper(
+            self.sparsity_coef_tau0, self.slab_shape_nu, self.slab_shape_s
+        )
+
+        # sample the parameters compute the predicted outputs.
+        beta = sample_reg_horseshoe(tau, c_sq, (n_targets, n_features))
+        mu = jnp.dot(x, beta.T)
+
+        # compute the likelihood.
+        sigma = numpyro.sample("sigma", Exponential(self.noise_hyper_lambda))
+        numpyro.sample("obs", Normal(mu, sigma), obs=y)
+
+    def _run_mcmc(self, x, y, **kwargs):
+        # set up a jax random key.
+        seed = kwargs.pop("seed", 0)
+        rng_key = random.PRNGKey(seed)
+        rng_key, rng_key_ = random.split(rng_key)
```

Comment:
It's a JAX-specific thing, please see here: [JAX - The Sharp Bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#random-numbers)

I'll have a look and see if I can make it more readable.

Reply:
I guess it is not strictly necessary, since there are no other seeded operations after `mcmc.run`.

Reply:
Thanks, that's a great link, I think i get it.  Since we're only consuming one key, and `random.split(0)` is just as random as `0`, could we just use the `rng_key = random.PRNGKey(seed)`?  line?

Reply:
Absolutely. I think this was a leftover from some early experiments. This is addressed in https://github.com/dynamicslab/pysindy/pull/440/commits/952b98932c8e4faaf5fbe7313768bd276f34a4c7